src/qibolab/_core/instruments/qrng.py

```diff
@@ -0,0 +1,151 @@
+from scipy.linalg import toeplitz
+from serial import Serial
+
+from .abstract import Instrument
+
+__all__ = ["QRNG"]
+
+
+def unpackbits(x: npt.NDArray, num_bits: int) -> npt.NDArray:
+    if np.issubdtype(x.dtype, np.floating):
+        raise ValueError("numpy data type needs to be int-like")
+    xshape = list(x.shape)
+    x = x.reshape([-1, 1])
+    mask = 2 ** np.arange(num_bits, dtype=x.dtype).reshape([1, num_bits])
+    return (x & mask).astype(bool).astype(int).reshape(xshape + [num_bits])
+
+
+def generate_toeplitz(input_bits: int, extraction_ratio: int) -> npt.NDArray:
+    """Generate a pseudo-random Toeplitz matrix of dimension ``(input_bits, extraction_ratio)``."""
+    while True:
+        c = np.mod(np.random.permutation(input_bits), 2)
+        r = np.mod(np.random.permutation(extraction_ratio), 2)
+        if np.sum(c) == 6 and np.sum(r) == (extraction_ratio // 2):
+            return toeplitz(c, r)
+
+
+def extractor(m1: npt.NDArray, input_bits: int, extraction_ratio: int) -> npt.NDArray:
+    """Extract uniformly distributed integers from the device samples."""
+    m2 = unpackbits(m1, input_bits)
+    m2 = np.flip(m2)
+
+    m3 = generate_toeplitz(input_bits, extraction_ratio)
+
+    m4 = np.matmul(m2, m3)
+    m4 = m4.astype("int16")
+    m4 = np.mod(m4, 2)
+    m4 = np.packbits(m4)
+    m4 = np.right_shift(m4, 8 - extraction_ratio)
+    return m4
+
+
+def upscale(samples: npt.NDArray, input_bits=4, output_bits=32) -> npt.NDArray:
+    """Increase size of random bit strings by concatenating them."""
+    assert output_bits > input_bits
+    assert output_bits % input_bits == 0
+
+    factor = output_bits // input_bits
+    assert len(samples) % factor == 0
+
+    n = len(samples) // factor
+    upscaled = np.zeros(n, dtype=int)
+    for i, s in enumerate(np.reshape(samples, (factor, n))):
+        upscaled += s << (input_bits * i)
+    return upscaled
+
+
+class QRNG(Instrument):
+    """Driver to sample numbers from a Quantum Random Number Generator (QRNG).
+
+    Note that if we are not connected to a physical QRNG device, this will
+    return pseudo-random generated numbers using ``numpy.random``.
+    """
+
+    address: str
+    baudrate: int = 115200
+
+    port: Optional[Serial] = None
+
+    samples_per_number: int = 4
+    """Number of bytes to read from the device to generate a number."""
+    raw_dimension: int = 12
+    """Number of bits of the raw sampled numbers following normal distribution."""
+    extracted_bits: int = 4
+    """Number of bits of the uniformly distributed extracted samples."""
+
+    def connect(self):
+        if self.port is None:
+            self.port = Serial(self.address, self.baudrate)
+
+    def disconnect(self):
+        if self.port is not None:
+            self.port.close()
+            self.port = None
+
+    def _read(self) -> Optional[int]:
+        num_str = ""
+        while len(num_str) < self.samples_per_number:
+            sample = self.port.read(1)
```

Comment:
Is it mandatory to actually read one byte at a time?

Reply:
Just to clarify, I would have implemented the whole acquisition with something like:
```py
        num_str = self.port.read(self.samples_per_number).decode()
```

Reply:
Several parts of the code, including this function, were copied from external scripts/tutorials on how to use the QRNG device. For your particular point, I agree and I would also try to do something similar, I would just prefer to have a better understanding of how it affects the results before starting to make changes. Also, probably worth mentioning that in the original code, `self.samples_per_number` is hard-coded to 4. Here I decided to make it a variable/instrument setting as I am not sure what is special about this 4.

In short, I am not sure about several things, so I decided to stick to existing code for a first iteration.